### PR TITLE
#150001988 Create sidebar, toggle and display contribution bookmarks

### PR DESCRIPTION
**Pull request by andela-aolasode**
#### What does this PR do?
Allows user to view sidebar and toggle the bookmarking of contributions and displays bookmark on the side bar.

#### Description of Task to be completed? 
- Create the sidebar component
- Create App Wide Menus on the sidebar
- Create bookmark section that display all user's bookmarks
- Create bookmark component that toggles the bookmarking of a contribution
- Restructure routing such that the Navbar and Sidebar appears on all pages
- Ensure Navbar and Sidebar are sticky such that scrolling of page content does not move them

#### How should this be manually tested?
After cloning the repo

- Run yarn install
- Run yarn start
- Access the landing page via localhost:3000 and click on the LOGIN TO CONTRIBUTE button.
- On the dashboard click on the hamburger icon to display the sidebar and its contents.
- Click the bookmark icon on each contribution card to toggle bookmark and see it appear and disappear on the sidebar.
- Scroll the page to see the Navbar and Sidebar remain in position
- Click the add contribution button to move another page to see the Navbar and Sidebar there.

#### Any background context you want to provide?
- There was no sidebar
- Contributions were not bookmarkable
- Navbar has to be included on each page

#### What are the relevant pivotal tracker stories?
[#150001966](https://www.pivotaltracker.com/story/show/150001966)
[#150001988](https://www.pivotaltracker.com/story/show/150001988)
[#150473913](https://www.pivotaltracker.com/story/show/150473913)

#### Screenshots (if appropriate)
image
image
image
image
screen shot 2017-08-27 at 10 52 33 am
screen shot 2017-08-27 at 10 53 11 am
screen shot 2017-08-27 at 10 53 41 am
screen shot 2017-08-27 at 11 26 51 am

### CODE REVIEW

**Comment by BolajiOlajide** [comment]
Hey man, modify your PR title to conform to the convention highlighted in the engineering Wiki.

**Comment by andela-kakpobome** [comment]
Fix your PR title

**Comment by andela-kakpobome** [comment]
Also, why do we have 3 stories in one? Reviewing 28 files takes considerably more effort than it does a 3rd of that

**Comment by andela-uofoegbu** [comment]
@andela-aolasode Hey, Quick question
I noticed the instructions card moves to the bottom of the contributions cards when the sidebar is open.
Is that the expected behaviour? I believe it should still maintain it's position only reduce width to fit the size of the container

File: src/actions/ContributionsAction.ts [hunk 1]
```diff
@@ -11,9 +13,29 @@ const loadContributionsSuccess = (contributions): interfaces.IContributionsActio
...
+    const state = getState();
+    const id = contribution.id;
+    if (isBookmarked) {
+      return http.delete(`${config.API_BASE_URL}/contributions/bookmarks/${id}`)
+        .then((contributions) => {
+          dispatch(loadContributions());
+          dispatch(loadBookmarks());
+        });
+    } else {
+      return http.post(`${config.API_BASE_URL}/contributions/${id}/bookmarks`, {resourceType: 'contribution'})
```
Comment (andela-uofoegbu) [inline src/actions/ContributionsAction.ts]: I noticed the endpoint for `post` is different from the `delete` endpoint
I thought the whole point of a toggle is to hit the same endpoint using a `PATCH` request or something similar.
 I think the logic here to check if it's been bookmarked before should be backend implementation.

File: src/api/MockBookmarksApi.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        modifiedBookmarks = bookmarks.filter((element) => {
+          return element.bookmark_id !== contributionID;
+        });
+        resolve(modifiedBookmarks);
+      });
+    });
+  }
+}
+
+export default MockBookmarksApi;
```
Comment (andela-uofoegbu) [inline src/api/MockBookmarksApi.ts]: It would be better to use the endpoints on the apiary docs

File: src/common/bookmark/bookmark.component.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    this.state = {
+      isBookMarked : false,
+    };
+  }
+
+  /**
+   * @memberOf Bookmark
+   */
+  public componentDidMount() {
+    this.setState({isBookMarked: this.props.isBookmarked});
```
Comment (andela-uofoegbu) [inline src/common/bookmark/bookmark.component.tsx]: I think this should occur right before the component mounts using a `componentWillMount`.
Using a `componentDidMount`, sets the state after the component has mounted and if for some reason the props are not received fast enough ... e.g. slow internet, The user will always see an unchecked bookmark button before it changes to the accurate state.

File: src/interfaces/contributions.d.ts [hunk 4]
```diff
@@ -88,12 +90,13 @@ interface IContributionCardState {
 }
 
 interface IContributionCardProps {
-  contribution: {
+  contribution?: {
```
Comment (andela-uofoegbu) [inline src/interfaces/contributions.d.ts]: Should `contribution` be optional?

File: src/pages/contributions/ContributionList.component.tsx [hunk 5]
```diff
@@ -6,16 +6,17 @@ import ContributionCard from './ContributionsCard';
  *
  * @class ContributionList
  * @param contributions
+ * @param handleBookmark
  * @returns {any}
  */
-const ContributionList = ({ contributions }) => {
+const ContributionList = ({ contributions, handleBookmark }) => {
```
Comment (andela-uofoegbu) [inline src/pages/contributions/ContributionList.component.tsx]: What does the `handleBookmark` function do and does it have to be passed to the `contributionList`?

File: src/actions/ContributionsAction.ts (hunk 1 above)
Comment (andela-aolasode) [reply src/actions/ContributionsAction.ts]: @andela-uofoegbu We had to implement different endpoints because the `bookmarks` can be deleted from the sidebar also but can't be bookmark from the sidebar which splits the toggle thing into two.
Also, I think the idea of the toggle is to have one component (in this case the bookmark icon) to add and delete bookmarks on the contribution card.

File: src/api/MockContributionApi.ts [hunk 6]
```diff
@@ -28,36 +28,82 @@ const contributions = [
...
   public static getAllContributions() {
     return new Promise((resolve, reject) => {
       setTimeout(() => {
-        resolve(Object.assign([], contributions));
+        resolve(defaultContributions);
+      });
+    });
+  }
+
+  public static bookMarkContribution(contributionID, contributions = []) {
```
Comment (andela-kakpobome) [inline src/api/MockContributionApi.ts]: ### Critical

This is deeply concerning because:
1. This implementation is >=90% the same as the `unBookmarkContribution`, why can't we have more abstraction here?
2. We are still using the Mockup APIs. how come? why can't we leverage the Apiary docs.

File: src/api/MockBookmarksApi.ts (hunk 2 above)
Comment (andela-aolasode) [reply src/api/MockBookmarksApi.ts]: Thanks. We will be removing the entire mock as we won't be needing it anymore

File: src/common/bookmark/bookmark.component.tsx (hunk 3 above)
Comment (andela-aolasode) [reply src/common/bookmark/bookmark.component.tsx]: Okay, thanks. We will be implementing that.

File: src/interfaces/contributions.d.ts (hunk 4 above)
Comment (andela-aolasode) [reply src/interfaces/contributions.d.ts]: Thank you. We did that while trying track while some tests are failing. we will revert it.

File: src/pages/contributions/ContributionList.component.tsx (hunk 5 above)
Comment (andela-aolasode) [reply src/pages/contributions/ContributionList.component.tsx]: We intend to use the `Bookmark` component for `paths` too and the `handleBookmark` function is to call the necessary action at each time which that of `paths` will differ from that of `contributions`

File: src/actions/BookmarksActions.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+// axios
+import http from '../helpers/http.service';
+
+const loadBookmarksSuccess = (bookmarks): interfaces.IBookmarkActionProps => {
+  return { type: types.LOAD_BOOKMARKS_SUCCESS, bookmarks };
+};
+
+export const loadBookmarks = () => {
+  return ((dispatch) => {
+    return http.get(`${config.API_BASE_URL}/contributions/bookmarks/my`)
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/actions/BookmarksActions.ts]: Yes!

File: src/actions/ContributionsAction.ts [hunk 8]
```diff
@@ -11,9 +13,29 @@ const loadContributionsSuccess = (contributions): interfaces.IContributionsActio
   return { type: types.LOAD_CONTRIBUTIONS_SUCCESS, contributions };
 };
 
+export const toggleBookmark = (userId, contribution, isBookmarked) => {
+  return ((dispatch, getState) => {
+    const state = getState();
+    const id = contribution.id;
+    if (isBookmarked) {
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/actions/ContributionsAction.ts]: Okay

File: src/api/MockContributionApi.ts (hunk 6 above)
Comment (andela-aolasode) [reply src/api/MockContributionApi.ts]: No! We will remove the entire mock file

File: src/common/navbar/NavBar.scss [hunk 9]
```diff
@@ -279,3 +286,7 @@
     }
   }
 }
+
+.reduce-padding {
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/common/navbar/NavBar.scss]: Thanks. We will rename it.

File: src/api/MockContributionApi.ts (hunk 6 above)
Comment (andela-aolasode) [reply src/api/MockContributionApi.ts]: Done

File: src/pages/bookmarks/BookmarksPage.component.tsx [hunk 10]
```diff
@@ -0,0 +1,79 @@
+import * as React from 'react';
```
Comment (andela-kakpobome) [inline src/pages/bookmarks/BookmarksPage.component.tsx]: ### Critical

The name of this component is grossly misleading,  it's not a page it's a list of `BookmarkedItems`

File: src/pages/bookmarks/BookmarksPage.component.tsx [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+import { render } from 'react-dom';
+import { Icon } from 'react-inline-svg-icon';
+import { connect } from 'react-redux';
+import TimeAgo from 'react-timeago';
+
+import * as bookmarksActions from '../../actions/BookmarksActions';
```
Comment (andela-kakpobome) [inline src/pages/bookmarks/BookmarksPage.component.tsx]: ### Nitpick

Add the `// xxx` line separators for the imports

File: src/pages/bookmarks/BookmarksPage.component.tsx [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+/**
+ * @class BookmarkComponent
+ * @extends {React.Component<any, any>}
+ */
+@connect((state) => {
+  return {
+    bookmarks: state.bookmarks,
+  };
+})
+
```
Comment (andela-kakpobome) [inline src/pages/bookmarks/BookmarksPage.component.tsx]: ### Critical

What's this space about?

File: src/pages/bookmarks/BookmarksPage.component.tsx [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+export class BookmarkComponent extends
```
Comment (andela-kakpobome) [inline src/pages/bookmarks/BookmarksPage.component.tsx]: ### Critical

Why is the file name different from the class name?

File: src/pages/bookmarks/BookmarksPage.component.tsx [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+   */
+  public componentWillMount() {
+    this.props.dispatch(bookmarksActions.loadBookmarks());
+  }
+
+  /**
+   * @param {object} nextProps
+   * @memberof BookmarksComponent
+   */
+  public componentWillReceiveProps(nextProps) {
```
Comment (andela-kakpobome) [inline src/pages/bookmarks/BookmarksPage.component.tsx]: ### Critical

Render should be the last method in the file

File: src/pages/contributions/Contribution.scss [hunk 15]
```diff
@@ -5,13 +5,17 @@
...
+}
+
 .list-container {
   width: 36.25rem;
   margin-right: 1rem;
 }
 
 .instructions {
-  flex: 1;
+flex: 1;
```
Comment (andela-kakpobome) [inline src/pages/contributions/Contribution.scss]: ### Critical

This is not properly indented.

File: src/pages/contributions/ContributionList.component.tsx (hunk 5 above)
Comment (andela-kakpobome) [reply src/pages/contributions/ContributionList.component.tsx]: Good point, I believe we'll implement a better decoupled system if we define the `handleBookmark` inside the `ContributionCard` component. So each card "knows" what to do when it's child `Bookmark` component is clicked

File: src/pages/contributions/ContributionPage.component.tsx [hunk 16]
```diff
@@ -18,7 +18,7 @@ import ContributionList from './ContributionList.component';
  */
 @connect((state) => {
   return {
-    contributions: state.contributions,
+    contributions: [...state.contributions],
```
Comment (andela-kakpobome) [inline src/pages/contributions/ContributionPage.component.tsx]: ### Critical

Why is this change necessary?

File: src/pages/contributions/ContributionPage.component.tsx [hunk 17]
```diff
@@ -96,28 +97,40 @@ interfaces.IContributionsPageState> {
...
   }
 
   /**
+   * Toggles bookmark on contribution
+   *
+   * @param {object} contributionId
+   * @param {boolean} isBookmarked
+   */
+  public handleBookmark = (contribution, isBookmarked): void => {
+    const userId = 1;
```
Comment (andela-kakpobome) [inline src/pages/contributions/ContributionPage.component.tsx]: ### Critical

1. Why are we sending the ID? The backend already gets that from the token.
2. We should be using the ES6 style of class method definition, as opposed to creating anonymous functions
3. There should be a linespace after this method.

File: src/pages/contributions/ContributionPage.component.tsx [hunk 18]
```diff
@@ -96,28 +97,40 @@ interfaces.IContributionsPageState> {
...
           <ContributionFilter
             filterContributions={this.filterContributions}
             sortContributions={this.sortContributions}
             searchContributions={this.searchContributions}
           />
         </div>
-        <div className="page-content">
-          <ContributionList contributions={this.state.filteredContributions} />
+        <div className={this.state.pageContainerStyle}>
+          <ContributionList contributions={this.state.filteredContributions} handleBookmark={this.handleBookmark} />
```
Comment (andela-kakpobome) [inline src/pages/contributions/ContributionPage.component.tsx]: ### Critical

Remember to take this out based on my comment above

File: src/pages/contributions/ContributionsCard.tsx [hunk 19]
```diff
@@ -1,5 +1,8 @@
 import * as React from 'react';
+
 import { Icon } from 'react-inline-svg-icon';
+
+import Bookmark from '../../common/bookmark/bookmark.component';
```
Comment (andela-kakpobome) [inline src/pages/contributions/ContributionsCard.tsx]: ### Critical

Shouldn't these go under the `// components` import section? Also change, `// component` to `components`

File: src/pages/dashboard/dashboard.component.tsx [hunk 20]
```diff
@@ -39,7 +42,14 @@ export class DashboardComponent extends React.Component<interfaces.IUserActionPr
     this.state = {
       activeUser: {},
       authorised: true,
+      menuStyle: 'menu',
+      navBarStyle: 'container',
+      pageContainerStyle: 'page-content',
+      pageContentPadding: 'container-pad-up',
+      sideBarStyle: 'side-bar hide',
     };
```
Comment (andela-kakpobome) [inline src/pages/dashboard/dashboard.component.tsx]: ### Nitpick

Leave a linespace before the next line

File: src/pages/contributions/ContributionPage.component.tsx [hunk 21]
```diff
@@ -96,28 +97,40 @@ interfaces.IContributionsPageState> {
    */
   public componentWillReceiveProps(nextProps) {
     const contributions = nextProps.contributions;
+    const pageContainerStyle = nextProps.pageContainerStyle;
     this.setState({
```
Comment (andela-kakpobome) [inline src/pages/contributions/ContributionPage.component.tsx]: ### Critical

Let's start to enforce anonymous function style state updates.

File: src/reducers/bookmarksReducers.ts [hunk 22]
```diff
@@ -0,0 +1,16 @@
+import initialState from './initialState';
```
Comment (andela-kakpobome) [inline src/reducers/bookmarksReducers.ts]: ### Critical

This file name should not be `pluralised`

File: src/routes/routes.tsx [hunk 23]
```diff
@@ -2,6 +2,7 @@ import * as React from 'react';
 import { Provider } from 'react-redux';
 import { BrowserRouter as Router, Link, Route, Switch } from 'react-router-dom';
 
+import ContributionPage from '../pages/contributions/ContributionPage.component';
```
Comment (andela-kakpobome) [inline src/routes/routes.tsx]: ### Nitpick

Add the `linespace import section header` here

File: src/common/bookmark/bookmark.component.tsx [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+    this.state = {
+      isBookMarked : false,
+    };
+  }
+
+  /**
+   * @memberOf Bookmark
+   */
+  public componentWillMount() {
+    this.setState({isBookMarked: this.props.isBookmarked});
```
[the remark this answers is not in the document]
Comment (andela-uofoegbu) [reply src/common/bookmark/bookmark.component.tsx]: @andela-kakpobome 
I had a similar scenario with my upvote task where i set the state in `componentDidMount` and for a brief moment when the page loads the vote count for all contributions shows as 0 (which was the initial state).

I think the only way that can be avoided would be to set the state right before the component mounts. That way it loads with the correct value

File: src/pages/contributions/ContributionList.component.tsx (hunk 5 above)
Comment (andela-aolasode) [reply src/pages/contributions/ContributionList.component.tsx]: Done

File: src/common/bookmark/bookmark.component.tsx [hunk 25]
```diff
@@ -0,0 +1,56 @@
+import * as React from 'react';
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/common/bookmark/bookmark.component.tsx]: Done

File: src/common/bookmark/bookmark.component.tsx [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * @class Bookmark
+ * @extends {React.Component<any, any>}
+ */
+export default class Bookmark extends React.Component<any, any> {
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/common/bookmark/bookmark.component.tsx]: Added!

File: src/common/bookmark/bookmark.component.tsx [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+  constructor(props) {
+    super(props);
+    this.state = {
+      isBookMarked : false,
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/common/bookmark/bookmark.component.tsx]: Done!

File: src/common/bookmark/bookmark.component.tsx [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+
+  /**
+   * @returns {object} bookmark icon
+   * @memberOf Bookmark
+   */
+  public render() {
+    return(
+      <div>
+        {this.state.isBookMarked ?
+          <a className="bookmark" onClick={() => {
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/common/bookmark/bookmark.component.tsx]: Done!

File: src/common/bookmark/bookmark.test.tsx [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+import { Provider, store } from 'react-redux';
+import { stub } from 'sinon';
+import configureStore from '../../store/store';
+
+import Bookmark from './bookmark.component';
+import { toggleBookmark } from '../../actions/ContributionsAction';
+
+const store = configureStore({});
+
+const bookmark = {"approvedAt": null,
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/common/bookmark/bookmark.test.tsx]: Done!

File: src/common/navbar/NavBar.component.tsx [hunk 30]
```diff
@@ -15,6 +15,7 @@ export class NavBar extends React.Component<any, any> {
     this.state = {
       picture: '',
     };
+    this.handleMenuClick = this.handleMenuClick.bind(this);
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/common/navbar/NavBar.component.tsx]: Done!

File: src/common/navbar/NavBar.component.tsx [hunk 31]
```diff
@@ -138,4 +129,8 @@ export class NavBar extends React.Component<any, any> {
       </div>
     );
   }
+
+  private handleMenuClick() {
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/common/navbar/NavBar.component.tsx]: Done!

File: src/common/navbar/NavBar.scss [hunk 32]
```diff
@@ -2,23 +2,40 @@
 
 .nav-bar {
   display: flex;
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/common/navbar/NavBar.scss]: Done!

File: src/common/navbar/NavBar.scss [hunk 33]
```diff
@@ -100,6 +107,7 @@
       justify-content: center;
       width: 1.5rem;
       height: 1.5rem;
+      border: 1px solid map-get($colors, primary);
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/common/navbar/NavBar.scss]: Okay

File: src/interfaces/bookmarks.d.ts [hunk 34]
```diff
@@ -0,0 +1,41 @@
+import * as types from '../types';
+
+/**
+ * @interface iBookmarks
+ */
+interface IBookmark {
+  bookmarkId: number;
+  type?: string;
+  path: string;
+  dateCreated?: string;
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/interfaces/bookmarks.d.ts]: Corrected

File: src/interfaces/bookmarks.d.ts [hunk 35]
```diff
@@ -0,0 +1,41 @@
...
+  userId?: number;
+  category: string;
+  contributionId?: number;
+  createdAt?: string;
+  status?: string;
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/interfaces/bookmarks.d.ts]: Corrected

File: src/interfaces/bookmarks.d.ts [hunk 36]
```diff
@@ -0,0 +1,41 @@
...
+ */
+interface IBookmarkActionProps {
+  bookmarks?: IBookmark;
+  type: types.LOAD_BOOKMARKS_SUCCESS;
+}
+
+/**
+ * @interface IBookmarksComponentState
+ */
+interface IBookmarksComponentState {
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/interfaces/bookmarks.d.ts]: Done

File: src/pages/bookmarks/Bookmarks.scss [hunk 37]
```diff
@@ -0,0 +1,64 @@
...
+    }
+  }
+
+  .bookmarked-contributions {
+    overflow: scroll;
+    height: 24rem;
+
+    .bookmark-title {
+      color: map_get($grey, 400);
+      font-size: 16px;
```
[the remark this answers is not in the document]
Comment (andela-aolasode) [reply src/pages/bookmarks/Bookmarks.scss]: Okay

File: src/pages/bookmarks/BookmarksPage.component.tsx (hunk 10 above)
Comment (andela-aolasode) [reply src/pages/bookmarks/BookmarksPage.component.tsx]: Done

File: src/pages/bookmarks/BookmarksPage.component.tsx (hunk 11 above)
Comment (andela-aolasode) [reply src/pages/bookmarks/BookmarksPage.component.tsx]: Done

File: src/pages/bookmarks/BookmarksPage.component.tsx (hunk 13 above)
Comment (andela-aolasode) [reply src/pages/bookmarks/BookmarksPage.component.tsx]: Corrected

File: src/pages/bookmarks/BookmarksPage.component.tsx (hunk 14 above)
Comment (andela-aolasode) [reply src/pages/bookmarks/BookmarksPage.component.tsx]: Done

File: src/pages/contributions/Contribution.scss (hunk 15 above)
Comment (andela-aolasode) [reply src/pages/contributions/Contribution.scss]: Corrected

File: src/pages/contributions/ContributionPage.component.tsx (hunk 16 above)
Comment (andela-aolasode) [reply src/pages/contributions/ContributionPage.component.tsx]: It's an oversight. It has been corrected

File: src/pages/contributions/ContributionPage.component.tsx (hunk 18 above)
Comment (andela-aolasode) [reply src/pages/contributions/ContributionPage.component.tsx]: Done

File: src/pages/contributions/ContributionsCard.tsx (hunk 19 above)
Comment (andela-aolasode) [reply src/pages/contributions/ContributionsCard.tsx]: Done

File: src/pages/contributions/ContributionPage.test.tsx [hunk 38]
```diff
@@ -21,11 +21,8 @@ const store = configureStore({});
 
 describe('Contributions Test', () => {
   describe('Contributions Card', () => {
-    const wrapper: any = mount(
-      <Provider store={store}>
-        <ContributionCard contribution={sampleContribution} />
-      </Provider>,
-    );
+    const wrapper: any = mount(<Provider store={store}><ContributionCard
+       contribution={ sampleContribution } /></Provider>);
```
Comment (andela-kakpobome) [inline src/pages/contributions/ContributionPage.test.tsx]: ### Nitpick

Why are there spaces on either side of the curly braces?

File: src/pages/bookmarks/BookmarkList.component.tsx [hunk 39]
```diff
@@ -0,0 +1,85 @@
...
+  return {
+    bookmarks: state.bookmarks,
+  };
+})
+export class BookmarkList extends
+ React.Component<interfaces.IBookmarkComponentProps, interfaces.IBookmarksComponentState>  {
+  constructor(props: interfaces.IBookmarkComponentProps) {
+    super(props);
+    this.state = {
+      bookmarks: [],
```
Comment (andela-kakpobome) [inline src/pages/bookmarks/BookmarkList.component.tsx]: ### Critical

This should be  of type `IBookmarkItem[]`

File: src/pages/contributions/ContributionsCard.tsx [hunk 40]
```diff
@@ -48,14 +53,30 @@ const getCategoryIcon = (cardType) => {
  * @param contribution
  * @returns {any}
  */
+@connect((state) => {
+   return {
+     contributions: state.contributions,
+   };
+ })
+
```
Comment (andela-kakpobome) [inline src/pages/contributions/ContributionsCard.tsx]: ### Critical

Remove this unnecessary line

File: src/reducers/initialState.ts [hunk 41]
```diff
@@ -1,5 +1,6 @@
 export default {
   activeUser: {},
+  bookmarks: [],
```
Comment (andela-kakpobome) [inline src/reducers/initialState.ts]: ### Critical

Let's make this explicitly of the `IBookmarkItem []`, and in a similar vain, for contributions

File: src/interfaces/bookmarks.d.ts [hunk 42]
```diff
@@ -0,0 +1,45 @@
...
+  type: string;
+}
+
+/**
+ * @interface IBookmarkComponentProps
+ */
+interface IBookmarkComponentProps {
+  bookmarks?: IBookmarkItem;
+  loadBookmarks?: () => void;
+  dispatch?: (x) => void;
```
Comment (andela-kakpobome) [inline src/interfaces/bookmarks.d.ts]: ### Critical

Are these really supposed to be optional?

File: src/actions/ContributionsAction.ts [hunk 43]
```diff
@@ -11,9 +13,52 @@ const loadContributionsSuccess = (contributions): interfaces.IContributionsActio
   return { type: types.LOAD_CONTRIBUTIONS_SUCCESS, contributions };
 };
 
+export const toggleBookmark = (contribution, isBookmarked) => {
```
Comment (andela-kakpobome) [inline src/actions/ContributionsAction.ts]: ### Critical

These concern bookmarks, and hence should be in the lit of bookmark actions

File: src/actions/BookmarksActions.ts [hunk 44]
```diff
@@ -0,0 +1,43 @@
...
+    return element.contributionId !== contributionId;
+  });
+
+  return ((dispatch) => {
+    dispatch(loadBookmarksSuccess(modifiedBookmarks));
+  });
+};
+
+export const addBookmark = (bookmarks, contribution, createdAt) => {
+  const newBookmark = {
```
Comment (andela-kakpobome) [inline src/actions/BookmarksActions.ts]: ### Critical

This should be of type `IBookmarkItem[]`

File: src/actions/ContributionsAction.ts [hunk 45]
```diff
@@ -11,9 +13,52 @@ const loadContributionsSuccess = (contributions): interfaces.IContributionsActio
...
+  return ((dispatch, getState) => {
+    const state = getState();
+    const contributions = state.contributions;
+    const bookmarks = state.bookmarks;
+    const id = contribution.id;
+
+    if (isBookmarked) {
+      dispatch(toggleContributionBookmark(id, contributions));
```
Comment (andela-kakpobome) [inline src/actions/ContributionsAction.ts]: ### Critical

Since we call this in both the `if & else` sections, let it be dispatched before the conditional, so we avoid the repitition

File: src/reducers/initialState.ts (hunk 41 above)
Comment (andela-aolasode) [reply src/reducers/initialState.ts]: We are actually setting the value of the `initialState` not declaring a variable.

Comment (andela-kakpobome) [reply src/reducers/initialState.ts]: Lol, you're right, my bad, as you were!
